lib/ueberauth/strategy/cognito.ex

```diff
@@ -54,21 +65,37 @@ defmodule Ueberauth.Strategy.Cognito do
   defp exchange_code_for_token(%Plug.Conn{params: %{"code" => code}} = conn) do
     config = Config.get_config()
 
-    with {:ok, token} <- request_token(conn, code, config),
-         {:ok, jwks} <- request_jwks(config),
+    with {:ok, token} <- request_token(conn, code, config) do
+      extract_and_verify_token(conn, token, config)
+    else
+      {:error, :cannot_fetch_tokens} ->
+        set_errors!(conn, error("aws_response", "Non-200 error code from AWS"))
+    end
+  end
+
+  defp exchange_code_for_token(conn) do
+    set_errors!(conn, error("no_code", "Missing code param"))
+  end
+
+  defp extract_and_verify_token(conn, token, config) do
+    with {:ok, jwks} <- request_jwks(config),
          {:ok, id_token} <-
            config.jwt_verifier.verify(
              token["id_token"],
              jwks,
              config
            ) do
-      conn
-      |> put_private(:cognito_token, token)
-      |> put_private(:cognito_id_token, id_token)
-    else
-      {:error, :cannot_fetch_tokens} ->
-        set_errors!(conn, error("aws_response", "Non-200 error code from AWS"))
+      conn =
+        conn
+        |> put_private(:cognito_token, token)
+        |> put_private(:cognito_id_token, id_token)
 
+      if token["refresh_token"] do
+        put_private(conn, :cognito_refresh_token, token["refresh_token"])
```

Comment:
If you know `token` is a map (and I think we do) it's better to use `Map.get` rather than the `Access` protocol:
```elixir
if refresh_token = Map.get(token, "refresh_token") do
  put_private(conn, :cognito_refresh_token, refresh_token)
else
  conn
end
```

Reply:
For that matter, what's the reasoning behind putting it in the conn separately? I don't see it being used anywhere.

The private part of the conn is just scratch space for us to use in `credentials/1`, `uid/1`, `info/1`, and `extra/1`, and should be deleted in the `handle_cleanup!/1` callback.

The refresh token already exists in `token` and we already return it in `credentials/1` as a key in the `%Ueberauth.Auth.Credentials{}` struct.

Reply:
Now that you bring it up, I don't think we do need this--it's left over from another approach I tried.

Reply:
👍 . In that case, you don't need to `conn = ` in the expression above, either.